src/alda/sound.clj

```diff
@@ -251,35 +241,83 @@
                    events)
          (sort-by :offset))))
 
-(defn schedule-event!
-  [^SynthesisEngine engine offset f]
-  (let [ts  (TimeStamp. offset)
-        cmd (proxy [ScheduledCommand] [] (run [] (f)))]
-    (.scheduleCommand engine ts cmd)))
-
-(defn schedule-events!
-  [events score playing? wait]
+(defn score-to-sequence
+  [events score]
   (let [{:keys [instruments audio-context]} score
-        engine (:synthesis-engine @audio-context)
-        begin  (.getCurrentTime ^SynthesisEngine engine)
-        end!   #(deliver wait :done)]
-    (pdoseq-block [{:keys [offset instrument duration] :as event} events]
-      (let [inst   (-> instrument instruments)
-            start! #(when @playing?
-                      (if-let [f (:function event)]
-                        (future (f))
-                        (start-event! audio-context event inst)))
-            stop!  #(when-not (:function event)
-                      (stop-event! audio-context event inst))]
-        (schedule-event! engine (+ begin
-                                   (/ offset 1000.0)) start!)
-        (when-not (:function event)
-          (schedule-event! engine (+ begin
-                                     (/ offset 1000.0)
-                                     (/ duration 1000.0)) stop!))))
-    (schedule-event! engine (+ begin
-                               (/ (score-length events) 1000.0)
-                               1) end!)))
+        {:keys [midi-synth midi-channels]} @audio-context
+        channels  (.getChannels ^Synthesizer midi-synth)
+
+        ;; TODO make resolution configurable
+        seq (new Sequence Sequence/PPQ Sequence/SMPTE_24)
```

Comment:
Showing my math: SMTPE_24 at a resolution 1 would be 24 ticks per second (24 frames per second, 1 tick per frame). At 60 bpm (1 quarter note beat per second), timing could be accurate up to 24 notes per beat, which I think is way more granular than any composer would ever need. You can fit sixteen 64th notes into the span of a quarter note, and you rarely ever see 64th notes in written scores, and if you do, it's going to be at a very slow tempo, almost certainly slower than 60 bpm because 32nd notes are already very fast at 60 bpm. And that's just 64th notes. 24 notes per beat is the equivalent of 96th notes!